walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,116 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Iterator;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    long duration = request.getDuration();
+    Collection<String> attendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    ArrayList<TimeRange> mandatoryTimes = new ArrayList<>();
+    ArrayList<TimeRange> optionalTimes = new ArrayList<>();
+
+    if (duration > 24 * 60) {
+      return mandatoryTimes;
+    }
+    mandatoryTimes.add(TimeRange.WHOLE_DAY);
+    optionalTimes.add(TimeRange.WHOLE_DAY);
+    if ((attendees.isEmpty() && optionalAttendees.isEmpty()) || events.isEmpty()) {
+      return mandatoryTimes;
+    }
+
+    // For each event, if there are no event attendees that are in the request, skip it.
+    // If the event has any regular attendees, process the event in mandatoryTimes, otherwise
+    // the event only has optional attendees, so it will be processed in optionalTimes
+    for (Event event : events) {
+      Set eventAttendees = event.getAttendees();
+      boolean attendingEvent = false;
+      for (String attendee : attendees) {
+        if (eventAttendees.contains(attendee)) {
+          attendingEvent = true;
+          break;
+        }
+      }
+      boolean optionals = false;
+      if (attendingEvent == false) {
+        for (String attendee : optionalAttendees) {
+          if (eventAttendees.contains(attendee)) {
+            optionals = true;
+            break;
+          }
+        }
```

Comment:
Since this is equivalent to the loop above, except for the list of attendees, think about making a single util function that returns true or false.